packages/firestore/src/core/query.ts

```diff
@@ -447,57 +447,79 @@ export abstract class Filter {
   /**
    * Creates a filter based on the provided arguments.
    */
-  static create(field: FieldPath, op: RelationOp, value: FieldValue): Filter {
-    if (value.isEqual(NullValue.INSTANCE)) {
-      if (op !== RelationOp.EQUAL) {
+  static create(field: FieldPath, op: Operator, value: FieldValue): Filter {
+    if (field.isKeyField()) {
+      assert(
+        value instanceof RefValue,
+        'Comparing on key, but filter value not a RefValue'
+      );
+      assert(
+        op !== Operator.ARRAY_CONTAINS &&
+          op !== Operator.ARRAY_CONTAINS_ANY &&
+          op !== Operator.IN,
+        `'${op.toString()}' queries don't make sense on document keys.`
+      );
+      return new KeyFieldFilter(field, op, value as RefValue);
+    } else if (value.isEqual(NullValue.INSTANCE)) {
+      if (op !== Operator.EQUAL) {
         throw new FirestoreError(
           Code.INVALID_ARGUMENT,
           'Invalid query. You can only perform equals comparisons on null.'
         );
       }
       return new NullFilter(field);
     } else if (value.isEqual(DoubleValue.NAN)) {
-      if (op !== RelationOp.EQUAL) {
+      if (op !== Operator.EQUAL) {
         throw new FirestoreError(
           Code.INVALID_ARGUMENT,
           'Invalid query. You can only perform equals comparisons on NaN.'
         );
       }
       return new NanFilter(field);
+    } else if (op === Operator.ARRAY_CONTAINS) {
+      return new ArrayContainsFilter(field, value);
+    } else if (op === Operator.IN) {
+      assert(
+        value instanceof ArrayValue,
+        'IN filter has invalid value: ' + value.toString()
+      );
+      return new InFilter(field, value as ArrayValue);
+    } else if (op === Operator.ARRAY_CONTAINS_ANY) {
+      return new ArrayContainsAnyFilter(field, value as ArrayValue);
```

Comment:
We should presumably have the same `value instanceof ArrayValue` assert here (as above).

Reply:
Yeah.  I don't entirely disagree but when the cast happens so far from the validation, it's hard to know it's safe... so I don't really have a better answer... Might end up cleaner if we moved some of the create() logic into the public API or something.